### ISSUE TITLE
Install Win10 on virtio-blk disk will fail.

### ISSUE THREAD

**Issue by x2c3z4** [spdk/spdk#2330]
# Bug report
Windows ISO: cn_windows_10_business_editions_version_20h2_updated_april_2021_x64_dvd_0a0fc490
Install Win10 on virtio-blk disk will fail.

This is the log from the vhost-user backend.
```
vhost-blk.test.1s0: max IOVs in request reached (req_idx = 1), iovcnt(129 == 129)
We have set the virtio-blk pci config 'max_seg' to 126, but the virtio-blk guest drive sends the io which's iovcnt is 129.
```
Stable version 0.1.185 looks ok.

More information is here:
https://github.com/virtio-win/kvm-guest-drivers-windows/issues/578

## Current Behavior
Install Windows 10 fail.

## Possible Solution
Using virtio-win Stable version 0.1.185.

## Steps to Reproduce
1. Install Windows 10;
2. Mount virtio-win 0.1.190 iso and install the virtio driver;
3. Install will fail.

**Comment by x2c3z4** [comment]
@changpe1  Do you have any suggestions about this issue?
I have fired this bug to virtio-win half a year ago, and I read the virtio-win code, there is no any evidence.
So I think we should fix this issue in SPDK/DPDK.

Maybe support a larger max_seg from 126 to (512-2) is a good choice.

**Comment by changpe1** [comment]
> max IOVs in request reached

Looks like Windows doesn't follow that configuration parameter, you may change that parameter to see it can work with Windows or not.

**Comment by x2c3z4** [comment]
> > max IOVs in request reached
> 
> Looks like Windows doesn't follow that configuration parameter, you may change that parameter to see it can work with Windows or not.

Yes, I have tested that the max iov count is 514 when installing win10 using the latest virtio-win, so if we set the SPDK_VHOST_IOVS_MAX to 515, it works.
The total iov is:
0:        16(header)
1-512: 512*4k(data)
513:    1(status)

**Comment by changpe1** [comment]
Set SPDK_VHOST_IOVS_MAX to 515 will waste a lot of memory, e.g: if your backend devices are Intel NVMe SSDs, the maximum transfer data size is only 128KiB, we need to split the request in the bdev layer into smaller ones.  A workaround maybe to set this value can be configured via RPC, users know the exact environment they are using, when using with Windows environment, set it to a big number.

**Comment by x2c3z4** [comment]
> Set SPDK_VHOST_IOVS_MAX to 515 will waste a lot of memory, e.g: if your backend devices are Intel NVMe SSDs, the maximum transfer data size is only 128KiB, we need to split the request in the bdev layer into smaller ones. 

I think the memory usage is in control.
One vhost blk task's iovs memory: 515 * sizeof(struct iovec) = 515*16 = 8240
Set iodepth is 128, one vhost blk controller's iovs memory = 1054720.
The data memory in iovs is from guest OS, so they don't increase the spdk/dpdk memory.
Even if the bdev layer splits the IO, I think no memory overhead is added.

> A workaround maybe to set this value can be configured via RPC, users know the exact environment they are using, when using with Windows environment, set it to a big number.

I think adding a rpc to configure it for Windows is a bad idea, because introducing this configuration will cause maintenance difficulties.

**Comment by changpe1** [comment]
Have you considered that the issue may not occur in vhost but may occur in the bdev layer if we just increase this value greater than 512?  The bdev layer will also consume resources to do split IO.

**Comment by x2c3z4** [comment]
> Have you considered that the issue may not occur in vhost but may occur in the bdev layer if we just increase this value greater than 512? The bdev layer will also consume resources to do split IO.

As long as the data is not copied, the resource consumption is relatively small.

**Comment by changpe1** [comment]
> > Have you considered that the issue may not occur in vhost but may occur in the bdev layer if we just increase this value greater than 512? The bdev layer will also consume resources to do split IO.
> 
> As long as the data is not copied, the resource consumption is relatively small.

we do have a similar issue here #2284, timeout happened due to limited resources in bdev layer.
